### PR TITLE
Update default build to gcc-10

### LINKED ISSUE #11138: Add gcc-10 to CI.
It was released and now we have to start support it.

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now ClickHouse uses gcc-10 for the release build. Fixes #11138.

### CODE REVIEW

File: cmake/warnings.cmake [hunk 1]
```diff
@@ -23,7 +23,7 @@ option (WEVERYTHING "Enables -Weverything option with some exceptions. This is i
 # Control maximum size of stack frames. It can be important if the code is run in fibers with small stack size.
 # Only in release build because debug has too large stack frames.
 if ((NOT CMAKE_BUILD_TYPE_UC STREQUAL "DEBUG") AND (NOT SANITIZE))
-    add_warning(frame-larger-than=16384)
+    add_warning(frame-larger-than=32768)
```
Comment (alesapin) [inline cmake/warnings.cmake:26]: Just check that we don't have any other errors.

**Comment by alesapin** [comment]
Still have 
```
FAILED: src/CMakeFiles/dbms.dir/Storages/MergeTree/MergeTreePartition.cpp.o 
/usr/bin/ccache /usr/bin/g++-10  -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=7 -DAWS_SDK_VERSION_PATCH=231 -DBUNDLED_STATIC_JEMALLOC=1 -DCONFIGDIR=\"\" -DENABLE_OPENSSL_ENCRYPTION -DHAVE_CONFIG_H -DLIBSASL_EXPORTS=1 -DLZ4_DISABLE_DEPRECATE_WARNINGS=1 -DOBSOLETE_CRAM_ATTR=1 -DOBSOLETE_DIGEST_ATTR=1 -DPLUGINDIR=\"\" -DPOCO_ENABLE_CPP11 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DSASLAUTHD_CONF_FILE_DEFAULT=\"\" -DSNAPPY_CODEC_AVAILABLE -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUNALIGNED_OK -DUSE_FASTMEMCPY=1 -DUSE_JEMALLOC=1 -DUSE_REPLXX=1 -DUSE_XXHASH=1 -DWITH_COVERAGE=0 -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -I../contrib/sentry-native/include -Iincludes/configs -I../src -Isrc -Isrc/Core/include -Icontrib/cyrus-sasl-cmake -I../contrib/krb5/src/include -Icontrib/krb5-cmake/include -I../base/common/.. -Ibase/common/.. -I../contrib/cityhash102/include -I../contrib/FastMemcpy -I../contrib/cctz/include -Icontrib/zlib-ng -I../contrib/zlib-ng -I../base/pcg-random/. -I../contrib/aws-c-common/include -I../contrib/aws-c-event-stream/include -Icontrib/aws-s3-cmake/include -I../contrib/lz4/lib -I../base/mysqlxx/.. -I../contrib/mariadb-connector-c/include -Icontrib/mariadb-connector-c/include -I../contrib/zstd/lib -I../contrib/libuv/include -isystem ../contrib/orc/c++/include -isystem contrib/orc/c++/include -isystem ../contrib/protobuf/src -isystem ../contrib/sparsehash-c11 -isystem ../contrib/openldap-cmake/linux_x86_64/include -isystem ../contrib/openldap/include -isystem ../contrib/avro/lang/c++/include -isystem ../contrib/pdqsort -isystem ../contrib/llvm/llvm/include -isystem contrib/llvm/llvm/include -isystem ../contrib/libcxx/include -isystem ../contrib/libcxxabi/include -isystem ../contrib/libunwind/include -isystem ../contrib/jemalloc-cmake/include -isystem ../contrib/libdivide/. -isystem ../contrib/cppkafka/include -isystem ../contrib/librdkafka-cmake/include -isystem ../contrib/librdkafka/src -isystem contrib/librdkafka-cmake/auxdir -isystem ../contrib/AMQP-CPP/include -isystem ../contrib/openssl-cmake/linux_x86_64/include -isystem ../contrib/openssl/include -isystem ../contrib/libbtrie/include -isystem ../contrib/boost -isystem ../contrib/msgpack-c/include -isystem ../contrib/aws/aws-cpp-sdk-s3/include -isystem ../contrib/aws/aws-cpp-sdk-core/include -isystem ../contrib/libhdfs3/include -isystem ../contrib/re2 -isystem ../contrib/poco/Net/include -isystem ../contrib/poco/Foundation/include -isystem ../contrib/poco/NetSSL_OpenSSL/include -isystem ../contrib/poco/Crypto/include -isystem ../contrib/poco/Util/include -isystem ../contrib/poco/JSON/include -isystem ../contrib/poco/XML/include -isystem ../contrib/replxx/include -isystem ../contrib/fmtlib-cmake/../fmtlib/include -isystem ../contrib/double-conversion -isystem ../contrib/ryu -isystem contrib/re2_st -isystem ../contrib/croaring -isystem ../contrib/poco/MongoDB/include -isystem ../contrib/icu/icu4c/source/i18n -isystem ../contrib/icu/icu4c/source/common -isystem ../contrib/capnproto/c++/src -isystem ../contrib/arrow/cpp/src -isystem ../contrib/arrow-cmake/cpp/src -isystem ../contrib/avro/lang/c++/api -isystem ../contrib/libc-headers/x86_64-linux-gnu -isystem ../contrib/libc-headers -fdiagnostics-color=always -std=c++2a -fsized-deallocation  -pipe -msse4.1 -msse4.2 -mpopcnt   -Wall  -Werror -Wextra -Wframe-larger-than=32768 -O2 -g -DNDEBUG -O3  -fno-pie   -D OS_LINUX -fno-tree-loop-distribute-patterns -Wbool-compare -Wcast-align -Wcast-qual -Wdelete-incomplete -Wdisabled-optimization -Wduplicated-cond -Wenum-compare -Winit-self -Wlogical-not-parentheses -Wlogical-op -Wmaybe-uninitialized -Wmisleading-indentation -Wmissing-include-dirs -Wnon-virtual-dtor -Wno-return-local-addr -Wodr -Wold-style-cast -Wplacement-new=2 -Wpointer-arith -Wredundant-decls -Wreorder -Wshadow -Wshift-negative-value -Wsized-deallocation -Wsizeof-array-argument -Wsizeof-pointer-memaccess -Wsuggest-override -Wswitch-bool -Wtautological-compare -Wtrampolines -Wunused -Wvector-operation-performance -Wno-sequence-point -nostdinc++ -MD -MT src/CMakeFiles/dbms.dir/Storages/MergeTree/MergeTreePartition.cpp.o -MF src/CMakeFiles/dbms.dir/Storages/MergeTree/MergeTreePartition.cpp.o.d -o src/CMakeFiles/dbms.dir/Storages/MergeTree/MergeTreePartition.cpp.o -c ../src/Storages/MergeTree/MergeTreePartition.cpp
In file included from ../contrib/libc-headers/string.h:494,
                 from ../contrib/libcxx/include/string.h:60,
                 from ../contrib/libcxx/include/cstring:60,
                 from ../contrib/libcxx/include/algorithm:641,
                 from ../contrib/libcxx/include/__bit_reference:15,
                 from ../contrib/libcxx/include/vector:274,
                 from ../src/Core/Row.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.cpp:1:
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'void writeHexByteLowercase(UInt8, void*)' at ../src/Common/hex.h:39:11,
    inlined from 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const' at ../src/Storages/MergeTree/MergeTreePartition.cpp:85:30:
../contrib/libc-headers/x86_64-linux-gnu/bits/string_fortified.h:34:33: error: writing 2 bytes into a region of size 1 [-Werror=stringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/Storages/MergeTree/MergeTreePartition.cpp: In member function 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const':
../src/Storages/MergeTree/MergeTreePartition.cpp:54:12: note: at offset 92 to object 'result' with size 24 declared here
   54 |     String result;
      |            ^~~~~~
In file included from ../contrib/libc-headers/string.h:494,
                 from ../contrib/libcxx/include/string.h:60,
                 from ../contrib/libcxx/include/cstring:60,
                 from ../contrib/libcxx/include/algorithm:641,
                 from ../contrib/libcxx/include/__bit_reference:15,
                 from ../contrib/libcxx/include/vector:274,
                 from ../src/Core/Row.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.cpp:1:
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'void writeHexByteLowercase(UInt8, void*)' at ../src/Common/hex.h:39:11,
    inlined from 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const' at ../src/Storages/MergeTree/MergeTreePartition.cpp:85:30:
../contrib/libc-headers/x86_64-linux-gnu/bits/string_fortified.h:34:33: error: writing 2 bytes into a region of size 0 [-Werror=stringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/Storages/MergeTree/MergeTreePartition.cpp: In member function 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const':
../src/Storages/MergeTree/MergeTreePartition.cpp:54:12: note: at offset 92 to object 'result' with size 24 declared here
   54 |     String result;
      |            ^~~~~~
In file included from ../contrib/libc-headers/string.h:494,
                 from ../contrib/libcxx/include/string.h:60,
                 from ../contrib/libcxx/include/cstring:60,
                 from ../contrib/libcxx/include/algorithm:641,
                 from ../contrib/libcxx/include/__bit_reference:15,
                 from ../contrib/libcxx/include/vector:274,
                 from ../src/Core/Row.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.cpp:1:
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'void writeHexByteLowercase(UInt8, void*)' at ../src/Common/hex.h:39:11,
    inlined from 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const' at ../src/Storages/MergeTree/MergeTreePartition.cpp:85:30:
../contrib/libc-headers/x86_64-linux-gnu/bits/string_fortified.h:34:33: error: writing 2 bytes into a region of size 0 [-Werror=stringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/Storages/MergeTree/MergeTreePartition.cpp: In member function 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const':
../src/Storages/MergeTree/MergeTreePartition.cpp:54:12: note: at offset 92 to object 'result' with size 24 declared here
   54 |     String result;
      |            ^~~~~~
In file included from ../contrib/libc-headers/string.h:494,
                 from ../contrib/libcxx/include/string.h:60,
                 from ../contrib/libcxx/include/cstring:60,
                 from ../contrib/libcxx/include/algorithm:641,
                 from ../contrib/libcxx/include/__bit_reference:15,
                 from ../contrib/libcxx/include/vector:274,
                 from ../src/Core/Row.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.h:3,
                 from ../src/Storages/MergeTree/MergeTreePartition.cpp:1:
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'void writeHexByteLowercase(UInt8, void*)' at ../src/Common/hex.h:39:11,
    inlined from 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const' at ../src/Storages/MergeTree/MergeTreePartition.cpp:85:30:
../contrib/libc-headers/x86_64-linux-gnu/bits/string_fortified.h:34:33: error: writing 2 bytes into a region of size 0 [-Werror=stringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/Storages/MergeTree/MergeTreePartition.cpp: In member function 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const':
../src/Storages/MergeTree/MergeTreePartition.cpp:54:12: note: at offset 92 to object 'result' with size 24 declared here
   54 |     String result;
      |            ^~~~~~
cc1plus: all warnings being treated as errors
```
https://github.com/ClickHouse/ClickHouse/issues/10508
Trying to disable this warning for this code.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
> /ClickHouse/src/Storages/MergeTree/MergeTreePartition.cpp:33:32: error: unknown warning group '-Wstringop-overflow'

Should suppress warning only for new gcc.

**Comment by alexey-milovidov** [comment]
Surprisingly, it didn't help.

**Comment by alesapin** [comment]
yarr
```
FAILED: src/CMakeFiles/unit_tests_dbms.dir/Core/tests/gtest_multienum.cpp.o 
/usr/bin/ccache /usr/bin/g++-10  -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=7 -DAWS_SDK_VERSION_PATCH=231 -DENABLE_OPENSSL_ENCRYPTION -DGTEST_HAS_POSIX_RE=0 -DPOCO_ENABLE_CPP11 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUNALIGNED_OK -DUSE_FASTMEMCPY=1 -DUSE_REPLXX=1 -DWITH_COVERAGE=0 -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -I../contrib/sentry-native/include -Iincludes/configs -I../contrib/cityhash102/include -I../contrib/libfarmhash -I../src -Isrc -Isrc/Core/include -I../base/common/.. -Ibase/common/.. -I../contrib/FastMemcpy -I../contrib/cctz/include -Icontrib/zlib-ng -I../contrib/zlib-ng -I../base/pcg-random/. -I../contrib/aws-c-common/include -I../contrib/aws-c-event-stream/include -Icontrib/aws-s3-cmake/include -I../contrib/consistent-hashing -I../contrib/consistent-hashing-sumbur -I../base/mysqlxx/.. -I../contrib/mariadb-connector-c/include -Icontrib/mariadb-connector-c/include -I../contrib/libuv/include -I../contrib/libmetrohash/src -I../contrib/murmurhash/include -isystem ../contrib/libcxx/include -isystem ../contrib/libcxxabi/include -isystem ../contrib/libunwind/include -isystem ../contrib/googletest/googletest/include -isystem ../contrib/googletest/googletest -isystem ../contrib/llvm/llvm/include -isystem contrib/llvm/llvm/include -isystem ../contrib/base64 -isystem ../contrib/fastops -isystem ../contrib/msgpack-c/include -isystem ../contrib/aws/aws-cpp-sdk-s3/include -isystem ../contrib/aws/aws-cpp-sdk-core/include -isystem ../contrib/libhdfs3/include -isystem ../contrib/re2 -isystem ../contrib/boost -isystem ../contrib/poco/Net/include -isystem ../contrib/poco/Foundation/include -isystem ../contrib/poco/NetSSL_OpenSSL/include -isystem ../contrib/poco/Crypto/include -isystem ../contrib/openssl-cmake/linux_x86_64/include -isystem ../contrib/openssl/include -isystem ../contrib/poco/Util/include -isystem ../contrib/poco/JSON/include -isystem ../contrib/poco/XML/include -isystem ../contrib/replxx/include -isystem ../contrib/fmtlib-cmake/../fmtlib/include -isystem ../contrib/double-conversion -isystem ../contrib/ryu -isystem contrib/re2_st -isystem ../contrib/croaring -isystem ../contrib/orc/c++/include -isystem contrib/orc/c++/include -isystem ../contrib/pdqsort -isystem ../contrib/AMQP-CPP/include -isystem ../contrib/libc-headers/x86_64-linux-gnu -isystem ../contrib/libc-headers -fdiagnostics-color=always -std=c++2a -fsized-deallocation  -pipe -msse4.1 -msse4.2 -mpopcnt   -Wall  -Werror -Wextra -Wframe-larger-than=32768 -O2 -g -DNDEBUG -O3  -fno-pie   -D OS_LINUX -fno-tree-loop-distribute-patterns -Wbool-compare -Wcast-align -Wcast-qual -Wdelete-incomplete -Wdisabled-optimization -Wduplicated-cond -Wenum-compare -Winit-self -Wlogical-not-parentheses -Wlogical-op -Wmaybe-uninitialized -Wmisleading-indentation -Wmissing-include-dirs -Wnon-virtual-dtor -Wno-return-local-addr -Wodr -Wold-style-cast -Wplacement-new=2 -Wpointer-arith -Wredundant-decls -Wreorder -Wshadow -Wshift-negative-value -Wsized-deallocation -Wsizeof-array-argument -Wsizeof-pointer-memaccess -Wsuggest-override -Wswitch-bool -Wtautological-compare -Wtrampolines -Wunused -Wvector-operation-performance -Wno-sequence-point -Wno-stringop-overflow -Wno-zero-as-null-pointer-constant -Wno-undef -Wno-sign-compare -Wno-used-but-marked-unused -Wno-missing-noreturn -Wno-gnu-zero-variadic-macro-arguments -nostdinc++ -pthread -MD -MT src/CMakeFiles/unit_tests_dbms.dir/Core/tests/gtest_multienum.cpp.o -MF src/CMakeFiles/unit_tests_dbms.dir/Core/tests/gtest_multienum.cpp.o.d -o src/CMakeFiles/unit_tests_dbms.dir/Core/tests/gtest_multienum.cpp.o -c ../src/Core/tests/gtest_multienum.cpp
In file included from ../src/Core/tests/gtest_multienum.cpp:5:
../src/Core/MultiEnum.h: In instantiation of 'bool operator==(ValueType, MultiEnum<EnumTypeT, StorageTypeT>) [with ValueType = int; <template-parameter-2-2> = void; EnumTypeT = {anonymous}::TestEnum; StorageTypeT = char8_t]':
../contrib/googletest/googletest/include/gtest/gtest.h:1544:11:   required from 'testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = MultiEnum<{anonymous}::TestEnum, char8_t>]'
../contrib/googletest/googletest/include/gtest/gtest.h:1571:23:   required from 'static testing::AssertionResult testing::internal::EqHelper::Compare(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = MultiEnum<{anonymous}::TestEnum, char8_t>; typename std::__1::enable_if<((! std::__1::is_integral<_Tp>::value) || (! std::__1::is_pointer<_Deleter>::value))>::type* <anonymous> = 0]'
../src/Core/tests/gtest_multienum.cpp:27:5:   required from here
../src/Core/MultiEnum.h:86:22: error: in C++20 this comparison calls the current function recursively with reversed arguments [-Werror]
   86 |         return right == left;
      |                ~~~~~~^~~~~~~
../src/Core/MultiEnum.h: In instantiation of 'bool operator==(ValueType, MultiEnum<EnumTypeT, StorageTypeT>) [with ValueType = int; <template-parameter-2-2> = void; EnumTypeT = {anonymous}::TestEnum; StorageTypeT = long unsigned int]':
../contrib/googletest/googletest/include/gtest/gtest.h:1544:11:   required from 'testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = MultiEnum<{anonymous}::TestEnum>]'
../contrib/googletest/googletest/include/gtest/gtest.h:1571:23:   required from 'static testing::AssertionResult testing::internal::EqHelper::Compare(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = MultiEnum<{anonymous}::TestEnum>; typename std::__1::enable_if<((! std::__1::is_integral<_Tp>::value) || (! std::__1::is_pointer<_Deleter>::value))>::type* <anonymous> = 0]'
../src/Core/tests/gtest_multienum.cpp:67:5:   required from here
../src/Core/MultiEnum.h:86:22: error: in C++20 this comparison calls the current function recursively with reversed arguments [-Werror]
cc1plus: note: unrecognized command-line option '-Wno-gnu-zero-variadic-macro-arguments' may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option '-Wno-used-but-marked-unused' may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
```

**Comment by alesapin** [comment]
Waiting for performance tests

File: cmake/warnings.cmake (hunk 1 above)
Comment (alesapin) [reply cmake/warnings.cmake]: Removed this

File: cmake/warnings.cmake [hunk 2]
```diff
@@ -169,9 +169,16 @@ elseif (COMPILER_GCC)
...
     if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 10)
+        # XXX: gcc10 stuck with this option while compiling GatherUtils code
+        # (anyway there are builds with clang, that will warn)
         add_cxx_compile_options(-Wno-sequence-point)
+        # XXX: gcc10 false positive with this warning in MergeTreePartition.cpp
+        #     inlined from 'void writeHexByteLowercase(UInt8, void*)' at ../src/Common/hex.h:39:11,
+        #     inlined from 'DB::String DB::MergeTreePartition::getID(const DB::Block&) const' at ../src/Storages/MergeTree/MergeTreePartition.cpp:85:30:
+        #     ../contrib/libc-headers/x86_64-linux-gnu/bits/string_fortified.h:34:33: error: writing 2 bytes into a region of size 0 [-Werror=stringop-overflow=]
+        #     34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
+        # For some reason (bug in gcc?) macro 'GCC diagnostic ignored "-Wstringop-overflow"' doesn't help.
```
Comment (azat) [inline cmake/warnings.cmake:181]: Indeed, looks like this one - https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94335#c7

**Comment by alesapin** [comment]
00652_replicated_mutations_zookeeper flaky, reported in CI Fest https://github.com/ClickHouse/ClickHouse/issues/14414.

**Comment by bobrik** [comment]
Is it expected to be backported into v20.9?

**Comment by alexey-milovidov** [comment]
It's not a bugfix.
Even more, there is no performance improvements in gcc-10 compared to gcc-9 on our codebase.

**Pull request merged**